libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -417,48 +477,101 @@ class kv_table {
       }
 
       bool operator==(const iterator& b) const {
-         return compare(b) == 0;
+         return base_iterator::compare(b) == 0;
       }
 
       bool operator!=(const iterator& b) const {
-         return compare(b) != 0;
+         return base_iterator::compare(b) != 0;
       }
 
       bool operator<(const iterator& b) const {
-         return compare(b) < 0;
+         return base_iterator::compare(b) < 0;
       }
 
       bool operator<=(const iterator& b) const {
-         return compare(b) <= 0;
+         return base_iterator::compare(b) <= 0;
       }
 
       bool operator>(const iterator& b) const {
-         return compare(b) > 0;
+         return base_iterator::compare(b) > 0;
       }
 
       bool operator>=(const iterator& b) const {
-         return compare(b) >= 0;
+         return base_iterator::compare(b) >= 0;
       }
+   };
 
-   private:
-      uint32_t itr;
-      status itr_stat;
+   class reverse_iterator : public base_iterator {
+      using base_iterator::itr;
+      using base_iterator::itr_stat;
+      using base_iterator::index;
 
-      const kv_index* index;
+   public:
+      using status = typename base_iterator::status;
 
-      int compare(const iterator& b) const {
-         bool a_is_end = !itr || itr_stat == status::iterator_end;
-         bool b_is_end = !b.itr || b.itr_stat == status::iterator_end;
-         if (a_is_end && b_is_end) {
-            return 0;
-         } else if (a_is_end && b.itr) {
+      reverse_iterator() = default;
+
+      reverse_iterator(uint32_t itr, status itr_stat, const kv_index* index) : base_iterator{itr, itr_stat, index} {}
+
+      reverse_iterator(reverse_iterator&& other) : base_iterator{std::move(other)} {}
```

Comment:
You might consider allowing a reverse_iterator to be constructed from an iterator.